src/dotty/tools/dotc/transform/Pickler.scala

```diff
@@ -40,9 +42,11 @@ class Pickler extends Phase {
 
     for { cls <- dropCompanionModuleClasses(topLevelClasses(unit.tpdTree))
           tree <- sliceTopLevel(unit.tpdTree, cls) } {
-      if (ctx.settings.YtestPickler.value) beforePickling(cls) = tree.show
       val pickler = new TastyPickler()
-      unit.picklers += (cls -> pickler)
+      if (ctx.settings.YtestPickler.value) {
+        beforePickling(cls) = tree.show
+        picklers(cls) = pickler
```

Comment:
I think this goes in the right direction, but in my heap dump I noticed that a massive amount of memory was taken by `beforePickling` so that may not be enough to get the tests to run smoothly. It may be helpful to do `beforePickling(cls) = null` and `picklers(cls) = null` in `testUnpickler` to free up these resources.